### ISSUE TITLE
Number of cores for large genomes

### ISSUE THREAD

**Issue by lyy005** [Gaius-Augustus/BRAKER#42]
Hi Katharina,
I have a question about setting the number of cores for the genome annotation. I have a eukaryotic genome (~500Mb). And it has a large number of scaffolds (39888 scaffolds). Most of the scaffolds are > 20Mb. And the rest of the scaffolds (14% of the genome) are fragmented. I set the --cores = 32 to annotate the genome in a fast way. Then I got this warning while running braker: 

file genome.fa contains a highly fragmented assembly (39888 scaffolds). This may lead to problems when running AUGUSTUS via braker in parallelized mode. You set --cores=32. You should run braker.pl in linear mode on such genomes, though (--cores=1). 

I wanted to make sure the annotation can be done in a relatively fast way. So I was wondering if there're other ways to speed up the annotation process? For example, can I split the genome into two parts, the longer scaffolds part and shorter scaffold part. And annotate the longer scaffolds (> 20Mb) with --cores=32 and shorter scaffold part with --cores=1? But I don't know if splitting the genome would also affect building the gene models. 

Thank you for any suggestions on this! 

Yiyuan

**Comment by KatharinaHoff** [comment]
Hi Yiyuan!

BRAKER will split a genome into one file per scaffold, and then run AUGUSTUS on all these scaffolds in parallel.

In case of highly fragmented assemblies, it is a better idea to split the genome into smaller files, but each file will still contain numerous scaffolds. BRAKER does not support that, at the moment, which is why it recommends that you skip parallelization all together.

I recommend the following:

a) look for the longest scaffolds, the part of the genome in which the majority of information resides. Use only that subset to run BRAKER.

b) if you need annotation for all the other tiny scaffolds, run AUGUSTUS outside of BRAKER with the BRAKER-trained AUGUSTUS parameters. You will find detailed parallelization instructions in https://math-inf.uni-greifswald.de/fileadmin/uni-greifswald/fakultaet/mnf/mathinf/stanke/augustus_wrp.pdf .

Best,

Katharina

**Comment by andreaswallberg** [comment]
Dear developers,

Sorry to open this topic but I think my question is better asked in this context rather starting from scratch with a new ticket.

I checked the braker.pl code, which says:

```
# unsure what is an appropriate limit, because it depends on the kernel and
# on the stack size. Use 30000 just to be sure. This will result in ~90000 files in the
# augustus_tmp folder.
if ( (scalar(@nScaffs) > 30000) && ($CPU > 1) ) {
    $prtStr = "#*********\n"
            . "# WARNING: in file " . __FILE__ ." at line ". __LINE__ ."\n"
            . "file $genome contains a highly fragmented assembly ("
            . scalar(@nScaffs)." scaffolds). This may lead "
            . "to problems when running AUGUSTUS via braker in parallelized "
            . "mode. You set --cores=$CPU. You should run braker.pl in linear "
            . "mode on such genomes, though (--cores=1).\n"
            . "#*********\n";
    print STDOUT $prtStr;
    print LOG $prtStr;
```

My question is:

For large/fragmented assemblies, is serial mode recommended due to operating system limitations (e.g. handling many files in a single directory) or accuracy in AUGUSTUS analyses (e.g. there is a clear benefit to gene recognition from running with --cores=1 compared to parallel execution)?

**Comment by KatharinaHoff** [comment]
The intended benefit of running in linear mode is that braker finishes
without crashing. It’s actually not an AUGUSTUS issue. AUGUSTUS itself will
run equally well on multiple cores (because AUGUSTUS always uses only one
core; braker splits data into many files to run AUGUSTUS tasks in parallel).

**Comment by KristinaGagalova** [comment]
Hi, 
Is there the option to stop BRAKER after it finishes training the Augustus parameters? I want to re-launch it on all the scaffolds later and I have a very large dataset (>100K scaffolds). Thank you in advance for the reply

**Comment by tomasbruna** [comment]
Hello,

there is no automatic stop, but you can stop BRAKER manually after the training of parameters finishes. This will not corrupt the parameter set. Just look for the following line in the log file and then you can kill the process.

```
#**********************************************************************************
#                               PREDICTING GENES WITH AUGUSTUS       
#**********************************************************************************
```

**Comment by splaisan** [comment]
HI all,
Sorry to reopen this old case but although I understood the logic, I hesitate for the implementation.

running on a 88cpu ubuntu server with 512GB RAM
I have a quite large genome of ~2GB and made of 38 pseudochromosomes or chr-arms larger than 20Mb (6 round 100Mb) and a myriad of leftover small contigs in the range 200bps to 10Kbps (>100'000).

Intuitively, I would like to

* run the process on 4 batches of 10 large chromosomes and give each serial job --threads 8
* run again with batches of the small contigs >500bps only with --threads=1
* merge the many GTF files into one final

does this sound good to you to speed up or should I split the big chromosomes further (8 jobs of 5 chr for instance) to give them more --treads (16)?

should I also limit the number of small contigs per batch to avoid a crash or is it OK with --threads=1 or can I just run all small contigs in one run? with --threads=1?

thanks for your advice

**Comment by KatharinaHoff** [comment]
Is it 88 or 8 cores? (Looks like 88 might be a typo, just want to be sure)

Stephane Plaisance ***@***.***> schrieb am Mo. 8. Mai 2023 um
16:07:

**Comment by splaisan** [comment]
No it is 88 and 512GB ram
Best
S

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by KatharinaHoff** [comment]
Thank you for clarifying. I would not split the genome at all. Possibly
remove short sequences, yes. You can run with 88 threads, not all will be
used efficiently. I usually go to only 72 or 48 threads, and even there
some threads won’t do anything for long times…

If it crashes, we can possibly advise on how to make it work…. Depends on
where it fails if it fails.

Let us know how it goes. I hope the genome is repeat masked…

Stephane Plaisance ***@***.***> schrieb am Mo. 8. Mai 2023 um
19:11:

**Comment by kinoppyi** [comment]
Hello developers,

I apologize for reopening this discussion, but I have a similar question. I am annotating the genome of a plant. The genome is large and divided into 18 scaffolds. Due to the large size of the genome, I decided to annotate it by splitting it into individual files and successfully annotated all the scaffolds.

At this point, I would like to proceed with functional annotation, but I have a question regarding the best way to merge the various files I have obtained (fasta for transcripts and proteins, and gtf) into a single file.

Let me explain better.

If I look at the files I obtained for the first chromosome (CH1) and the second (CH2), I see that the protein IDs are the same. For example:
CH1 has the following IDs in the braker.aa file:
`>g1.t1
>g2.t1
>g3.t1
etc...`

CH2 has the following IDs in the braker.aa file:
`>g1.t1
>g2.t1
>g3.t1
etc...`

They are the same, and this applies to all the braker.aa files I obtained for each chromosome.

Similarly, the gtf files have:
CH1
`Scaffold_00001  GeneMark.hmm3   gene          964686  964994  .  +  .  g1
Scaffold_00001  GeneMark.hmm3   transcript    964686  964994  .  +  .  g1.t1
Scaffold_00001  GeneMark.hmm3   CDS           964686  964994  .  +  0  transcript_id "g1.t1"; gene_id "g1";
Scaffold_00001  GeneMark.hmm3   exon          964686  964994  .  +  0  transcript_id "g1.t1"; gene_id "g1";
`
CH2
`Scaffold_00001  GeneMark.hmm3   gene          964686  964994  .  +  .  g1
Scaffold_00001  GeneMark.hmm3   transcript    964686  964994  .  +  .  g1.t1
Scaffold_00001  GeneMark.hmm3   CDS           964686  964994  .  +  0  transcript_id "g1.t1"; gene_id "g1";
Scaffold_00001  GeneMark.hmm3   exon          964686  964994  .  +  0  transcript_id "g1.t1"; gene_id "g1";
Scaffold_00002  AUGUSTUS        gene          15290   22269   .  -  .  g1
Scaffold_00002  AUGUSTUS        transcript    15290   22269   0.32  -  .  g1.t1
Scaffold_00002  AUGUSTUS        stop_codon    15290   15292   .  -  0  transcript_id "g1.t1"; gene_id "g1";
Scaffold_00002  AUGUSTUS        CDS           15290   15964   0.96  -  0  transcript_id "g1.t1"; gene_id "g1";
`

n this case, is it necessary to modify gene_id and transcript_id in the fasta files for proteins and transcripts, as well as in the gtf files, before merging all the files? If so, what is the best method to perform this operation?

Thank you for your help!
